### PR TITLE
Dev report sink implement cloud schema loader

### PR DESCRIPTION
**Pull request by manu-govind**
Change the report-sink service schema validation code to leverage that. See the Readme in the schema validation service on how to implement the same.

Health endpoint amended to report an issue if the folder with schema files can not be accessed.  Added to the dependencies section of the health rollup.

### CODE REVIEW

File: pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt [hunk 1]
```diff
@@ -1,14 +1,47 @@
...
 
 
+class SchemaLoaderConfiguration(environment: ApplicationEnvironment){
+    val schemaLoaderSystem = environment.config.tryGetString("ktor.schema_loader_system")?: ""
+    private val s3Bucket = environment.config.tryGetString("aws.s3.report_schema_bucket") ?: ""
+    private val s3Region = environment.config.tryGetString("aws.s3.report_schema_region") ?: ""
+    private val connectionString = environment.config.tryGetString("azure.blob_storage.connection_string") ?: ""
+    private val container = environment.config.tryGetString("azure.blob_storage.container") ?: ""
+
+        fun createSchemaLoader(): CloudSchemaLoader {
```
Comment (mkrystof) [inline pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt:21]: Please add missing function header.

File: pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt [hunk 2]
```diff
@@ -49,10 +84,11 @@ val RabbitMQPlugin = createApplicationPlugin(
 
     val factory = pluginConfig.getConnectionFactory()
     val queueName = pluginConfig.queue
-
+    //val schemaLoader: CloudSchemaLoader = pluginConfig.createSchemaLoader()
     lateinit var connection: Connection
     lateinit var channel: Channel
-
+    val environment: ApplicationEnvironment = this@createApplicationPlugin.application.environment
+    val schemaLoader = SchemaLoaderConfiguration(environment).createSchemaLoader() // Create the schema loader here
```
Comment (mkrystof) [inline pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt:91]: Really confused why this is mixed in with the rabbitmq plugin. Is this a mistake?

**Review by mkrystof — CHANGES_REQUESTED** [review]

File: libs/schema-validation/build.gradle.kts [hunk 3]
```diff
@@ -35,6 +35,9 @@ dependencies {
     implementation ("com.google.code.gson:gson:2.10.1")
     implementation ("com.sun.activation:javax.activation:1.2.0")
 
+    implementation(project(":libs:commons-database"))
```
Comment (mkrystof) [inline libs/schema-validation/build.gradle.kts:38]: What does the schema-validation need access to common-database for?  Can it be removed?

File: pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/Application.kt [hunk 4]
```diff
@@ -27,31 +32,51 @@ enum class MessageSystem {
...
-                    AzureServiceBusConfiguration(environment.config, configurationPath = "azure.service_bus") }
+                    AzureConfiguration(environment.config, configurationPath = "azure") }
 
             }
             MessageSystem.RABBITMQ.toString() -> {
                 single(createdAtStart = true) {
                     RabbitMQServiceConfiguration(environment.config, configurationPath = "rabbitMQ")
                 }
 
+                //For local and/or when the msg system is RabbitMQ -we need to access the cloud storage either blob or s3
```
[the remark this answers is not in the document]
Comment (manu-govind) [reply pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/Application.kt]: Will address this change in the PHD-367

File: pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/Application.kt [hunk 5]
```diff
@@ -79,14 +104,14 @@ fun Application.module() {
 
     when (messageSystem) {
         MessageSystem.AZURE_SERVICE_BUS -> {
-            serviceBusModule()
```
[the remark this answers is not in the document]
Comment (manu-govind) [reply pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/Application.kt]: Done

File: pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt [hunk 6]
```diff
@@ -1,14 +1,47 @@
...
 import gov.cdc.ocio.processingstatusapi.utils.SchemaValidation
+import gov.cdc.ocio.reportschemavalidator.loaders.CloudSchemaLoader
 import io.ktor.server.application.*
 import io.ktor.server.application.hooks.*
 import io.ktor.server.config.*
 import org.apache.qpid.proton.TimeoutException
 import java.io.IOException
 
 
+class SchemaLoaderConfiguration(environment: ApplicationEnvironment){
```
[the remark this answers is not in the document]
Comment (manu-govind) [reply pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt]: Done

File: pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/Application.kt [hunk 7]
```diff
@@ -17,6 +17,11 @@ enum class MessageSystem {
     RABBITMQ
 }
 
+enum class SchemaLoaderSystem {
```
[the remark this answers is not in the document]
Comment (manu-govind) [reply pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/Application.kt]: Will address that in PHDO-366

File: pstatus-report-sink-ktor/build.gradle [hunk 8]
```diff
@@ -74,6 +74,13 @@ dependencies {
     implementation 'org.danilopianini:khttp:1.3.1'
     implementation group: 'com.google.code.gson', name: 'gson', version: '2.8.9'
 
+    //Azure Blob
```
[the remark this answers is not in the document]
Comment (manu-govind) [reply pstatus-report-sink-ktor/build.gradle]: Done

File: pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt (hunk 1 above)
Comment (manu-govind) [reply pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt]: Done

File: pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt (hunk 2 above)
Comment (manu-govind) [reply pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/RabbitMQ.kt]: As discussed, will be addressing it in PHDO-367

File: libs/schema-validation/src/main/kotlin/gov/cdc/ocio/reportschemavalidator/utils/CloudSchemaLoaderConfiguration.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+import gov.cdc.ocio.reportschemavalidator.loaders.CloudSchemaLoader
+import io.ktor.server.application.*
+import io.ktor.server.config.*
+
+class CloudSchemaLoaderConfiguration(environment: ApplicationEnvironment){
+    private val schemaLoaderSystem = environment.config.tryGetString("ktor.schema_loader_system")?: ""
+    private val s3Bucket = environment.config.tryGetString("aws.s3.report_schema_bucket") ?: ""
+    private val s3Region = environment.config.tryGetString("aws.s3.report_schema_region") ?: ""
+    private val connectionString = environment.config.tryGetString("azure.blob_storage.connection_string") ?: ""
```
Comment (mkrystof) [inline libs/schema-validation/src/main/kotlin/gov/cdc/ocio/reportschemavalidator/utils/CloudSchemaLoaderConfiguration.kt:16]: Can we please rename this so it's obvious what the connection string is for?  The aws.s3.report_schema_bucket, etc...I like those names.  Can we rename to azure.blob_storage.report_schema_connection_string for example?  Same comment for the container name.

**Review by mkrystof — CHANGES_REQUESTED** [review]

File: pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/Application.kt [hunk 10]
```diff
@@ -79,14 +104,14 @@ fun Application.module() {
 
     when (messageSystem) {
         MessageSystem.AZURE_SERVICE_BUS -> {
-            serviceBusModule()
+            azureModule()
```
[the remark this answers is not in the document]
Comment (manu-govind) [reply pstatus-report-sink-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/Application.kt]: Done

File: libs/schema-validation/build.gradle.kts (hunk 3 above)
Comment (manu-govind) [reply libs/schema-validation/build.gradle.kts]: Not needed. Removed.

File: libs/schema-validation/src/main/kotlin/gov/cdc/ocio/reportschemavalidator/utils/CloudSchemaLoaderConfiguration.kt (hunk 9 above)
Comment (manu-govind) [reply libs/schema-validation/src/main/kotlin/gov/cdc/ocio/reportschemavalidator/utils/CloudSchemaLoaderConfiguration.kt]: Done

**Review by mkrystof — APPROVED** [review]

**Pull request merged**